File: src/domain/models/token.rs
use rand::{thread_rng, Rng};

const LENGTH: usize = 64;

#[derive(Clone, PartialEq)]
pub struct Token {
    raw: String,
}

impl Token {
    pub fn generate() -> Self {
        let mut rng = thread_rng();
        let mut raw = String::from("");

        for _ in 0..LENGTH {
            raw.push_str(&format!("{}", rng.gen_range(0..10)));
        }

        Self { raw }
    }
}

impl From<Token> for String {
    fn from(token: Token) -> Self {
        token.raw
    }
}

impl TryFrom<String> for Token {
    type Error = ();

    fn try_from(raw: String) -> Result<Self, Self::Error> {
        if raw.len() == LENGTH {
            Ok(Self { raw })
        } else {
            Err(())
        }
    }
}

File: src/domain/models/token.ts
import { Result } from "@utils";

export type Token = {
    _type: "token";
    raw: string;
};

const LENGTH = 64;

export const Token = {
    generate(): Token {
        const raw = Array(LENGTH)
            .fill(0)
            .map(() => Math.floor(Math.random() * 10))
            .join("");
        return { _type: "token", raw };
    },
    toString(token: Token): string {
        return token.raw;
    },
    tryFromString(raw: string): Result<Token, "not_a_token"> {
        const notAToken = raw.length !== LENGTH
            && Number.isNaN(Number(raw));
        if (notAToken) { return Result.err("not_a_token"); }
        return Result.ok({ _type: "token", raw });
    }
}
